src/Language/Haskell/TH/Instances.hs

```diff
@@ -140,6 +154,31 @@ instance Fail.MonadFail Q where
 # endif
 #endif
 
+#if !MIN_VERSION_template_haskell(2,17,0)
+instance Semi.Semigroup a => Semi.Semigroup (Q a) where
+  (<>) = liftA2 (Semi.<>)
+
+instance Monoid a => Monoid (Q a) where
+  mempty = return mempty
+#if MIN_VERSION_base(4,11,0)
+  mappend = (Semi.<>)
+#else
+  mappend = liftA2 mappend
+#endif
+
+instance MonadFix Q where
+  mfix k = do
+    m <- runIO newEmptyMVar
+    ans <- runIO (unsafeInterleaveIO (readMVar m
+#if MIN_VERSION_base(4,11,0)
```

Comment:
Hm. What happens if you run this code on pre-`4.11` versions of `base`?

Reply:
You get less obvious "where this came from?" `BlockedIndefinitelyOnMVar` exception.

I.e. the `catch` is there to help debugging. I don't think it's wise to try to recover from these exceptions. (I cannot think a good reason why you would do that).

Reply:
Ah, OK. Perhaps it is worth including a disclaimer that this will throw a different exception upon failure on old versions of `base`.

Reply:
Added.